README.md

```diff
@@ -1,7 +1,84 @@
 # rust-spiffe
-
 [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/maxlambrecht/rust-spiffe/blob/main/LICENSE)
 
-Rust client library implementation for SPIFFE, to interact with the Workload API to fetch and verify JWT and X.509 SVIDs.
+## Overview
+
+This library provides functions to interact with the [SPIFFE Workload API](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Workload_API.md) 
+to fetch X.509 and JWT SVIDs and Bundles. It also provides types that comply with the [SPIFFE standards](https://github.com/spiffe/spiffe/tree/main/standards).
 
 Under development.
+
+## Examples
+
+### Create a `WorkloadApiClient`
+
+Providing the endpoint socket path as parameter:
+
+```rust
+let client = WorkloadApiClient::new("unix:/tmp/spire-agent/api/public.sock")?;
+```
+
+Providing the endpoint socket path through the environment variable `SPIFFE_ENDPOINT_SOCKET`:
+
+```rust
+let client = WorkloadApiClient::default()?;
+```
+
+### Fetching X.509 materials
+
+```rust
+
+// fetch the default X.509 SVID
+let x509_svid: X509Svid = client.fetch_x509_svid()?;
+
+// fetch a set of X.509 bundles (X.509 public key authorities)
+let x509_bundles: X509BundleSet = client.fetch_x509_bundles()?;
+
+// fetch all the X.509 materials (SVIDs and bundles)
+let x509_context: X509Context = client.fetch_x509_context()?;
+
+// get the X.509 chain of certificates from the SVID
+let cert_chain: &Vec<Certificate> = x509_svid.cert_chain();
+
+// get the private key from the SVID
+let private_key: &PrivateKey = x509_svid.private_key();
+
+// parse a SPIFFE trust domain
+let trust_domain = TrustDomain::try_from("example.org")?;
+
+// get the X.509 bundle associated to the trust domain
+let x509_bundle: &X509Bundle = x509_bundles.get_bundle(&trust_domain).unwrap();
+
+// get the X.509 authorities (public keys) in the bundle
+let x509_authorities: &Vec<Certificate> = x509_bundle.authorities();
+```
+
+
+### Fetching JWT tokens and bundles and validating tokens
+
+```rust
+
+// parse a SPIFFE ID to ask a token for
+let spiffe_id = SpiffeId::try_from("spiffe://example.org/my-service")?;
+
+// fetch a jwt token for the provided SPIFFE-ID and with the target audience `service1.com`
+let jwt_token = client.fetch_jwt_token(&["audience1", "audience2"], Some(&spiffe_id))?;
```

Comment:
Maybe my SPIFFE knowledge is a bit rusted by now but why is the SPIFFE ID needed as a parameter here? Could it be the case that the workload does not know its SPIFFE ID if its set by another entity during registration?

edit: ah I see, this is an `Option` in case you want the `sub` claim to be populated. In our case, we mainly use the fetch method for the workload to *know* which is its SPIFFE ID

Reply:
The usual case would be to just pass `None` in the SPIFFE ID parameter and the Workload API will return a JWT token with 'sub' claim set with the SPIFFE ID corresponding to the default identity (the first one).

If the workload has multiple identities mapped to it, there is the possibility to select which SPIFFE ID to request the token for. 

In case it's requested a token with a SPIFFE ID that the workload it's not mapped with, the Workload API will return an empty response. 

Makes sense?

Reply:
That makes sense! Thank you for the explanation 👍